instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/TracingExecutionInterceptor.java

```diff
@@ -35,101 +42,121 @@
   static final ExecutionAttribute<Span> SPAN_ATTRIBUTE =
       new ExecutionAttribute<>("io.opentelemetry.auto.Span");
 
+  static final ExecutionAttribute<RequestType> REQUEST_TYPE_ATTRIBUTE =
+      new ExecutionAttribute<>("io.opentelemetry.auto.aws.RequestType");
+
   static final String COMPONENT_NAME = "java-aws-sdk";
 
-  private final Kind kind;
+  private final SpanFactory spanFactory;
+
+  private final Map<RequestType, SdkRequestDecorator> decoratorsMap = mapDecorators();
+  private final Map<String, String> fieldToAttribute = mapFieldToAttribute();
+
+  private Map<RequestType, SdkRequestDecorator> mapDecorators() {
+    Map<RequestType, SdkRequestDecorator> result = new EnumMap<>(RequestType.class);
+    result.put(RequestType.DynamoDB, new DbRequestDecorator());
+    return result;
+  }
+
+  private Map<String, String> mapFieldToAttribute() {
+    Map<String, String> result = new HashMap<>();
+    result.put("QueueUrl", "aws.queue.url");
+    result.put("Bucket", "aws.bucket.name");
+    result.put("QueueName", "aws.queue.name");
+    result.put("StreamName", "aws.stream.name");
+    result.put("TableName", "aws.table.name");
+    return result;
+  }
 
   TracingExecutionInterceptor(Kind kind) {
-    this.kind = kind;
+    this.spanFactory = new SpanFactory(kind);
+  }
+
+  private Optional<SdkRequestDecorator> decorator(ExecutionAttributes executionAttributes) {
```

Comment:
Let's just return the nullable value instead of `Optional`, we don't really use that pattern in this or the Java SDK repo. It could be a good time to add a dependency on checker framework annotations which we haven't done here yet for `@Nullable`

Reply:
I don't know if there is any performance difference, but other than that I would prefer `Optional`. But indeed, it should then be a global decision for the sake of consistency.

Reply:
There's a small performance difference - I think `Optional` can still make sense for public APIs where users may not have tooling to deal with `@Nullable` (though `Optional` works poorly with other JVM languages like Kotlin / Scala so still not great...), for our private methods, we should just do what we can at compile time at compile time :) https://github.com/uber/NullAway

Reply:
fwiw, I also prefer `null` + `@Nullable` + compile-time tooling https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/499#issuecomment-643560951

Reply:
Before we decide against Optionals everywhere besides module / component public API, there are few things we need to consider:
- it's already, at least in the AWS SDK interfacing part - eg `getValueForField` returns Optionals - ie inside AWS SDK tracing code it's consistent with the SDK itself 
- there is possible performance difference (invokedynamic for functional calls vs cmpz for plain-old NULL check) but IMO it's offset greatly by readability (no explicit branch) and additional safeguard (optional clearly says - beware, this might be null)
- I mostly agree with @iNikem comments regarding introduction of static type check - results might get ignored easily (if not failing the build upon a problem) - or will take a lot of effort to get the build running after introduction (if failing the build)
- out of my experience Optionals work better than static code checkers - for example if we break the checker (remove it accidentally, lib is broken etc) the code will become vulnerable, language construct will stay till it's manually removed / chaged
- removing Optionals now will be unsafe as null check is not implemented in the project yet

Reply:
We surely shouldn't add nullness checking without having it fail the build. I went ahead and sent #1203 finally to allow adding nullaway to modules.

> it's already, at least in the AWS SDK interfacing part - eg getValueForField returns Optionals - ie inside AWS SDK tracing code it's consistent with the SDK itself

It's more important for us to have consistency within our codebase than following the library patterns, unless there is a very specific use case. For example, I think our codebase becomes less consistent / maintainable for its authors if we prefer `Optional` in aws sdk instrumentation and `Nullable` in armeria instrumentation because of the library choices

> it's offset greatly by readability (no explicit branch) and additional safeguard (optional clearly says - beware, this might be null)

Readability argument is personal opinion I guess, but I find a small if statement to generally be at least as readable as a lambda. I don't think there's a difference in safeguardness when we have the checker in place (IntelliJ is excellent at highlighting usages of Nullable without check too)

> for example if we break the checker 

I feel this is true of all our tooling - if we break shadowJar configuration, agent doesn't work, etc. I've had great experience with static checking myself. So I'd like to be confident in our tooling as a whole :-D

Reply:
I merged #1203 in case you want to use it :) Otherwise I'm happy to add the checking later. But would be nice to stick with nullable here, I think that PR also shows the overwhelming use of it already in our codebase